language/changes/7-packages.md

```diff
@@ -0,0 +1,323 @@
+it wishes. Intuitively you can think of this as parametrizing the package
+`ExamplePkg` by the named address `NamedAddr`, and the package can then be
+instantiated later on by an importing package.
+
+`NamedAddr` can also be declared as:
+
+```
+[package]
+name = "ExamplePkg"
+...
+[addresses]
+NamedAddr = "0xCAFE"
+```
+
+which states that the named address `NamedAddr` is exactly `0xCAFE` and cannot be
+changed. This is is useful so other importing packages can use this named
+address without needing to worry about the exact value assigned to it.
+
+With these two different declaration methods, there are two ways that
+information about named addresses can flow in the package graph:
+* The former ("unassigned named addresses") allows named address values to flow
+  from the importation site to the declaration site.
+* The latter ("assigned named addresses") allows named address values to flow
+  from the declaration site upwards in the package graph to usage sites.
+
+With these two methods for flowing named address information throughout the
+package graph the rules around scoping and renaming become important to
+understand.
+
+## Scoping and Renaming of Named Addresses
+
+A named address `N` in a package `P` is in scope if:
+1. It declares a named address `N`; or
+2. A package in one of `P`'s transitive dependencies declares the named address
+  `N` and there is a dependency path in the package graph between between `P` and the
+  declaring package of `N` with no renaming of `N`.
+
+Additionally, every named address in a package is exported. Because of this and
+the above scoping rules each package can be viewed as coming with a set of
+named addresses that will be brought into scope when the package is imported,
+e.g., if the `ExamplePkg` package was imported, that importation would bring
+into scope the `NamedAddr` named address. Because of this, if `P` imports two
+packages `P1` and `P2` both of which declare a named address `N` an issue
+arises in `P`: which "`N`" is meant when `N` is referred to in `P`? The one
+from `P1` or `P2`? To prevent this ambiguity around which package a named
+address is coming from, we enforce that the sets of scopes introduced by all
+dependencies in a package are disjoint, and provide a way to _rename named
+addresses_ when the package that brings them into scope is imported.
+
+Renaming a named address when importing can be done as follows in our `P`,
+`P1`, and `P2` example above:
+
+```
+[package]
+name = "P"
+...
+[dependencies]
+P1 = { local = "some_path_to_P1", addr_subst = { "P1N" = "N" } }
+P2 = { local = "some_path_to_P2"  }
+```
+
+With this renaming `N` refers to the `N` from `P2` and `P1N` will refer to `N`
+coming from `P1`:
+
+```
+module N::A {
+    public fun x(): address { @P1N }
+}
+```
+
+It is important to note that _renaming is not local_: once a named address `N`
+has been renamed to `N'` in a package `P` all packages that import `P` will not
```

Comment:
Can you actually use `N'` as a package name? I'm guessing not, in which case it would be good to change the example to something that is a valid package name, e.g., `N2`